File: 21.py
# 21. Merge two sorted lists
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        cur1 = list1
        cur2 = list2
        list3 = None
        if cur1:
            if cur2:
                if cur1.val > cur2.val:
                    list3 = cur2
                    cur2 = cur2.next
                else: 
                    list3 = cur1
                    cur1 = cur1.next
            else:
                list3 = cur1
                cur1 = cur1.next
        else:
            if cur2:
                list3 = cur2
                cur2 = cur2.next
        cur3 = list3
        while cur1 or cur2:
            if cur1 and (not cur2):
                cur3.next = cur1
                cur1 = cur1.next
                cur3 = cur3.next
            elif (not cur1) and cur2:
                cur3.next = cur2
                cur2 = cur2.next
                cur3 = cur3.next
            else:
                if cur1.val > cur2.val:
                    cur3.next = cur2
                    cur2 = cur2.next
                    cur3 = cur3.next
                else:
                    cur3.next = cur1
                    cur1 = cur1.next
                    cur3 = cur3.next
        return list3

File: 21.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    // edge cases - both lists 0, list1 0, list2 0
    // else, set list1 to be lower, mark head
    // then, iterate through, if l1ptr.Val <= l2ptr.Val, iterate, else merge
    if list1 == nil {
        return list2
    }
    if list2 == nil {
        return list1
    }
    if list1.Val > list2.Val {
        list1, list2 = list2, list1
    }
    cur := &ListNode{
        Val: list1.Val,
    }
    head := cur
    list1 = list1.Next
    l1ptr, l2ptr := list1, list2
    for {
        if l1ptr == nil || l2ptr == nil {
            break
        }
        if l1ptr.Val <= l2ptr.Val {
            cur.Next = l1ptr
            l1ptr = l1ptr.Next
        } else {
            cur.Next = l2ptr
            l2ptr = l2ptr.Next
        }
        cur = cur.Next
    }
    if l1ptr != nil {
        cur.Next = l1ptr
    } else {
        cur.Next = l2ptr
    }
    return head
}
